### PR TITLE
Web takes optional tokenSwapURL and tokenRefreshURL for Authorization Code (without PKCE) and token swap

### PR DESCRIPTION
**Pull request by nzoschke**
Before, a Web Playback SDK developer could only use the internal in PKCE auth and refresh flow, which seems to always require an auth prompt.

Now, a developer can specify `tokenSwapURL` and `tokenRefreshURL`. If specified the library performs and Authorization Code (without PKCE) flow, and uses the swap and refresh URLs to exchange the code for a token, and to refresh the token respectively.

The token swap pattern is a first-class pattern on iOS, but the same concept applies to the web SDK for Spotify apps that manage access and refresh tokens with a server side component.

https://developer.spotify.com/documentation/ios/guides/token-swap-and-refresh/

### CODE REVIEW

**Comment by nzoschke** [comment]
I am testing this with something similar to the below. It appears to be working well with:

- Sending the user thru an Authorization Code (without PKCE) flow. For users that have already authenticated, this does not prompt them to auth again
- Exchanging the code for a token thru the tokenSwapURL
- Refreshing the token thru the tokenRefreshURL

```dart
  Future init() async {
    var apiURI = 'https://jukelab.herokuapp.com';
    var clientID = '<REDACTED>';
    var redirectURI = 'http://localhost:8686/callback.html';

    SpotifySdkPlugin.tokenSwapURL = '$apiURI/api/v1/spotify/token';
    SpotifySdkPlugin.tokenRefreshURL = '$apiURI/api/v1/spotify/refresh';

    var scope = [
      'app-remote-control',
      'playlist-modify-public',
      'playlist-read-collaborative',
      'playlist-read-private',
      'streaming',
      'user-library-modify',
      'user-library-read',
      'user-modify-playback-state',
      'user-read-currently-playing',
      'user-read-email',
      'user-read-playback-state',
      'user-read-private',
      'user-top-read',
    ].join(',');

    var token = await SpotifySdk.getAuthenticationToken(
      clientId: clientID,
      redirectUrl: redirectURI,
      scope: scope,
    );

    client = SpotifyClient(bc.BrowserClient(), token);

    var ok = await SpotifySdk.connectToSpotifyRemote(
      clientId: clientID,
      redirectUrl: redirectURI,
      playerName: 'JukeLab',
      scope: scope,
    );

    print('connect $ok');
  }
```

**Comment by nzoschke** [comment]
Something new to consider is that it would be nice if I could get the underlying access token to use for the Spotify API. 

`getAuthenticationToken` still returns a token which works, but I don't have a way to refresh it without duplicating and competing with the internal use of tokenRefreshURL.

Would you consider a `getAccessToken` helper?

**Comment by fotiDim** [comment]
Just a disclaimer, on iOS there `getAuthenticationToken` & `connectToSpotifyRemote` methods do the exact same thing under the hood only difference being that `getAuthenticationToken` actually returns the token. So in a nutshell on iOS, getAuthenticationToken = connectToSpotifyRemote + return token. On web it is very similar and frankly if it wasn't for Android we would have unified those 2 methods.

Now having said all that, I would like to understand the suggestion a bit better. Do you need a `getAccessToken` method that just returns the already generated access token without authenticating the user again?

Can you elaborate on `I don't have a way to refresh it without duplicating and competing with the internal use of tokenRefreshURL.`?

File: lib/spotify_sdk_web.dart [hunk 1]
```diff
@@ -83,6 +83,17 @@ class SpotifySdkPlugin {
...
+  static String? _tokenRefreshURL;
+
+  static set tokenSwapURL(String s) {
+    _tokenSwapURL = s;
+  }
+
+  static set tokenRefreshURL(String s) {
+    _tokenRefreshURL = s;
+  }
+
```
Comment (fotiDim) [inline lib/spotify_sdk_web.dart:96]: Why wouldn't we expose the getter?

File: lib/spotify_sdk_web.dart [hunk 2]
```diff
@@ -463,18 +497,38 @@ class SpotifySdkPlugin {
...
+      req = RequestOptions(
+        path: _tokenSwapURL!,
+        method: 'POST',
+        data: {
+          'code': parsedMessage.queryParameters['code'],
+          'redirect_uri': redirectUrl,
+        },
+        contentType: Headers.formUrlEncodedContentType,
+      );
+    }
```
Comment (fotiDim) [inline lib/spotify_sdk_web.dart:527]: RequestOptions req;
```suggestion
    if (_tokenSwapURL == null) {
      // build request to exchange auth code with PKCE for access and refresh tokens
    req = RequestOptions(
      path: 'https://accounts.spotify.com/api/token',
      method: 'POST',
      data: {
        'client_id': clientId,
        'grant_type': 'authorization_code',
        'code': parsedMessage.queryParameters['code'],
        'redirect_uri': redirectUrl,
        'code_verifier': codeVerifier
      },
      contentType: Headers.formUrlEncodedContentType,
    );
    }
    else {
        // or build request to exchange code with token swap
        // https://developer.spotify.com/documentation/ios/guides/token-swap-and-refresh/
      req = RequestOptions(
        path: _tokenSwapURL!,
        method: 'POST',
        data: {
          'code': parsedMessage.queryParameters['code'],
          'redirect_uri': redirectUrl,
        },
        contentType: Headers.formUrlEncodedContentType,
      );
    }
```

Let's use if-else here to make the 2 paths more explicit. I messed up the indentation and needs a couple of fixes for null safety but I hope you get the idea.

File: lib/spotify_sdk_web.dart [hunk 3]
```diff
@@ -492,15 +546,36 @@ class SpotifySdkPlugin {
...
+      req = RequestOptions(
+        path: _tokenRefreshURL!,
+        method: 'POST',
+        data: {
+          'refresh_token': refreshToken,
+        },
+        contentType: Headers.formUrlEncodedContentType,
+      );
+    }
+
```
Comment (fotiDim) [inline lib/spotify_sdk_web.dart:573]: Same as above

**Comment by nzoschke** [comment]
> Just a disclaimer, on iOS there getAuthenticationToken & connectToSpotifyRemote methods do the exact same thing under the hood only difference being that getAuthenticationToken actually returns the token....
> On web it is very similar

Thanks for the extra info about iOS. I don't see the similarity on web. Both `getAuthenticationToken` and `connectToSpotify` call `_authorizeSpotify under the hood, but `getAuthenticationToken` doesn't do anything to initialize the Web Playback player or anything.

> Now having said all that, I would like to understand the suggestion a bit better. Do you need a getAccessToken method that just returns the already generated access token without authenticating the user again?

Yes exactly. 

The use case is to use the `spotify` package or similar to make Web API calls in parallel with the SDK. 

I'd like to keep my web API client and the Web Playback SDK in sync and use the same accessToken and refresh mechanism. 

Otherwise there's yet another token flow for me to manage, and I fear one would refresh without the other knowing and things would get out of sync.

So for every API call I'd like to be able to use `SpotifySdk.accessToken` or similar. 

But thinking through that, then I'd also like some refresh mechanism, either `SpotifySdk.refreshToken()` to call if I get a 4XX from the API, or to get the accessToken thru an async getter that checks the expiry and refreshes it.

In a nutshell after this change and the web SDK has the access token and refresh built in, I'd like to reuse that to make Web API calls.

However I'm fine to punt on this and focus on this first bit just for the token swap.

File: lib/spotify_sdk_web.dart (hunk 1 above)
Comment (nzoschke) [reply lib/spotify_sdk_web.dart]: fixed.

File: lib/spotify_sdk_web.dart (hunk 2 above)
Comment (nzoschke) [reply lib/spotify_sdk_web.dart]: fixed

File: README.md [hunk 4]
```diff
@@ -95,6 +95,19 @@ On iOS Spotify starts playing music when attempting connection. This is a defaul
...
+
+```dart
+SpotifySdkPlugin.tokenSwapURL = 'https://example.com/api/spotify/token';
+SpotifySdkPlugin.tokenRefreshURL = 'https://example.com/api/spotify/refresh';
+````
+
+On web, this package will perform an Authorization Code (without PKCE) flow, then exchange the code and refresh the token with a backend service you run at the URLs provided.
+
+While the iOS SDK also supports the "token swap", this package does not yet configure the iOS SDK correctly.
+
```
Comment (fotiDim) [inline README.md:110]:
```suggestion

Token Swap is for now "web only". While the iOS SDK also supports the "token swap", this flow is not yet supported.

```

**Comment by fotiDim** [comment]
> Thanks for the extra info about iOS. I don't see the similarity on web. Both `getAuthenticationToken` and `connectToSpotify` call `_authorizeSpotify under the hood, but `getAuthenticationToken` doesn't do anything to initialize the Web Playback player or anything.
>
Let me put it this way, `getAuthenticationToken` and `connectToSpotify` are a bit overlapping on iOS and web. We could have gotten away by only having one of those two and going through the entire initialization. Then an additional getter for the access token would have been enough.

> I'd like to keep my web API client and the Web Playback SDK in sync and use the same accessToken and refresh mechanism.
> 
I do something similar and I leave this library do the refreshing and then update the accessToken on the `spotify` package.

> So for every API call I'd like to be able to use `SpotifySdk.accessToken` or similar.
> 
I agree that this is missing. My issue is that `getAuthenticationToken` semantically overlaps and it will create confusion. Let's discuss that in a new issue or PR.

> But thinking through that, then I'd also like some refresh mechanism, either `SpotifySdk.refreshToken()` to call if I get a 4XX from the API, or to get the accessToken thru an async getter that checks the expiry and refreshes it.
> 
Can be arranged. On the native iOS SDK there is the `reniewSession()` call that does exactly that.

**Review by fotiDim — APPROVED** [review]
